### ISSUE TITLE
Output embedding shape difference between exported onnx version vs. pt version of sentence-transformers model

### ISSUE THREAD

**Issue by yaman** [huggingface/optimum#1519]
### System Info

```shell
- `optimum` version: 1.14.0
- `transformers` version: 4.35.0
- Platform: Linux-6.5.8-arch1-1-x86_64-with-glibc2.35
- Python version: 3.11.5
- Huggingface_hub version: 0.17.3
- PyTorch version (GPU?): 2.1.0 (cuda availabe: True)
- Tensorflow version (GPU?): not installed (cuda availabe: NA)
```

### Who can help?

Hey there, @michaelbenayoun

I am converting sentence-transformers/clip-ViT-B-32-multilingual-v1 to onnx with following command;

```bash
optimum-cli export onnx -m sentence-transformers/clip-ViT-B-32-multilingual-v1 \
--task feature-extraction models/clip_vit_multilingual_onnx
```

with the following output;

```
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
Framework not specified. Using pt to export to ONNX.
Using the export variant default. Available variants are:
    - default: The default ONNX variant.
Using framework PyTorch: 2.1.0
[...transformers/models/distilbert/modeling_distilbert.py:223): TracerWarning: torch.tensor results are registered as constants in the trace. You can safely ignore this warning if you use this function to create tensors out of constant variables that would be the same every time you call this function. In any other case, this might cause the trace to be incorrect.
  mask, torch.tensor(torch.finfo(scores.dtype).min)
Post-processing the exported models...
Deduplicating shared (tied) weights...
Validating ONNX model models/clip_vit_multilingual_onnx/model.onnx...
	-[✓] ONNX model output names match reference model (last_hidden_state)
	- Validating ONNX Model output "last_hidden_state":
		-[✓] (2, 16, 768) matches (2, 16, 768)
		-[✓] all values close (atol: 0.0001)
The ONNX export succeeded and the exported model was saved at: models/clip_vit_multilingual_onnx
```

importing the code and running it;

```python
from transformers import AutoTokenizer
from optimum.onnxruntime import ORTModelForFeatureExtraction

tokenizer = AutoTokenizer.from_pretrained("models/clip_vit_multilingual_onnx")
model = ORTModelForFeatureExtraction.from_pretrained("models/clip_vit_multilingual_onnx")

inputs = tokenizer("What am I using?", "Using DistilBERT with ONNX Runtime!", return_tensors="pt")
outputs = model(**inputs)
last_hidden_state = outputs.last_hidden_state
list(last_hidden_state.shape)
```

gives the output slightly(export output: (2, 16, 768), actual: [1, 21, 768])  confirming the export command output:

```
[1, 21, 768]
```

If I use the same model from sentence-transformers:

```python
from sentence_transformers import SentenceTransformer, util
import torch

text_model = SentenceTransformer('sentence-transformers/clip-ViT-B-32-multilingual-v1')
texts = [
    "What am I using?", "Using DistilBERT with ONNX Runtime!", # Spanish: a beach with palm trees
]
text_embeddings = text_model.encode(texts)
print(text_embeddings.shape)
```

output: 

```
(2, 512)
```

as documented in: https://huggingface.co/sentence-transformers/clip-ViT-B-32-multilingual-v1;

```
SentenceTransformer(
  (0): Transformer({'max_seq_length': 128, 'do_lower_case': False}) with Transformer model: DistilBertModel 
  (1): Pooling({'word_embedding_dimension': 768, 'pooling_mode_cls_token': False, 'pooling_mode_mean_tokens': True, 'pooling_mode_max_tokens': False, 'pooling_mode_mean_sqrt_len_tokens': False})
  (2): Dense({'in_features': 768, 'out_features': 512, 'bias': False, 'activation_function': 'torch.nn.modules.linear.Identity'})
)
```

which is wildly different: (2, 512) vs (1, 21, 768).

Obviously resulting vector is different as well.

Am I doing something wrong?

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction (minimal, reproducible, runnable)

1-

```bash
optimum-cli export onnx -m sentence-transformers/clip-ViT-B-32-multilingual-v1 \
--task feature-extraction models/clip_vit_multilingual_onnx
```
2- 

```python
from transformers import AutoTokenizer
from optimum.onnxruntime import ORTModelForFeatureExtraction

tokenizer = AutoTokenizer.from_pretrained("models/clip_vit_multilingual_onnx")
model = ORTModelForFeatureExtraction.from_pretrained("models/clip_vit_multilingual_onnx")

inputs = tokenizer("What am I using?", "Using DistilBERT with ONNX Runtime!", return_tensors="pt")
outputs = model(**inputs)
last_hidden_state = outputs.last_hidden_state
list(last_hidden_state.shape)
```

3-

```python
from sentence_transformers import SentenceTransformer, util
import torch

text_model = SentenceTransformer('sentence-transformers/clip-ViT-B-32-multilingual-v1')
texts = [
    "What am I using?", "Using DistilBERT with ONNX Runtime!", # Spanish: a beach with palm trees
]
text_embeddings = text_model.encode(texts)
print(text_embeddings.shape)
```

### Expected behavior

shapes of both model encoded embeddings should be same;

```original(2, 512) != onnx_version(2, 16, 768)```

at least embedding dimensions should fit; `768` should be `512`

**Comment by yaman** [comment]
**Follow up:**

It seems sentence-transformers/clip-ViT-B-32-multilingual-v1 pytorch model **has pooling and dense layers**, which at some point converts the output format from dim 768 to 512. 

In case of exporting from **optimum-cli,** apparently  it ignores pooling and dense part, resulting an output shape of [1, 21, 768].

Looking at the optimum library, I can't seem to find any information about pooling and dense layer inclusion. optimum-cli will not obviously check 1_pooling or 2_dense folders at https://huggingface.co/sentence-transformers/clip-ViT-B-32-multilingual-v1/tree/main next to pytorch_model.bin 

I am looking for a way to combine 1_pooling and 2_dense layers into exported onnx. Not a clue yet other than creating these layers from scratch with pytorch(which is way beyond my abilities).

**Comment by mattweber** [comment]
@yaman any luck?  Running into the same issue.

**Comment by yaman** [comment]
> @yaman any luck?  Running into the same issue.

Are you having issues with clip-vit-multi as well?

**Comment by MadsLang** [comment]
> > @yaman any luck?  Running into the same issue.
> 
> Are you having issues with clip-vit-multi as well? If so, I successfully exported with a little help from torch.onnx, I can share fully working code to export.

I am having the same issue with `sentence-transformers/distiluse-base-multilingual-cased-v1`. If you could share some code, it would be greatly appreciated! :)

**Comment by mattweber** [comment]
@yaman  thank you!  I will give this a try with my model and see if I can get it working.

**Comment by fxmarty** [comment]
Thank you @mattweber @yaman

Basically Transformers behavior corresponds to outputting `token_embeddings`, not `sentence_embeddings`. I will fix to output both.

**Comment by fxmarty** [comment]
Fixed in https://github.com/huggingface/optimum/pull/1589.

Please give a try to Optimum from source and
```
optimum-cli export onnx -m sentence-transformers/clip-ViT-B-32-multilingual-v1 --task feature-extraction clip_vit_multilingual_onnx
```

**Comment by yaman** [comment]
I can confirm that output of encode contains both token and sentence embeddings and output sentence_embedding vector is close to original(atol-6). 

**Here is the test code:**

_install latest optimum from source_
```bash
python -m pip install git+https://github.com/huggingface/optimum.git
```

_test both original model and onnx output sentence_embedding vector_

```python
import torch
import numpy as np
import onnxruntime as ort
from transformers import AutoTokenizer
from sentence_transformers import SentenceTransformer


model = SentenceTransformer('sentence-transformers/clip-ViT-B-32-multilingual-v1')
tokenizer = AutoTokenizer.from_pretrained('sentence-transformers/clip-ViT-B-32-multilingual-v1')

# Prepare the input
text = "texts are close."
inputs = tokenizer(text, padding='longest', truncation=True, max_length=128, return_tensors='pt')

# Run the PyTorch model
pytorch_output =  model.encode(text, convert_to_tensor=True, device='cpu')

# Convert the inputs to numpy arrays for the ONNX model
inputs_onnx = {name: tensor.numpy() for name, tensor in inputs.items()}

# Run the ONNX model
sess = ort.InferenceSession("clip_vit_multilingual_onnx/model.onnx")
onnx_output = sess.run(None, inputs_onnx)

# Compare the outputs
print("Are the outputs close?", np.allclose(pytorch_output.detach().numpy(), onnx_output[1], atol=1e-6))


# Calculate the differences between the outputs
differences = pytorch_output.detach().numpy() - onnx_output[1]

# Print the standard deviation of the differences
print("Standard deviation of the differences:", np.std(differences))

print("pytorch_output size:", pytorch_output.size())
print("onnx_output size:", onnx_output[1].shape)
```

**result:**

```
Are the outputs close? True
Standard deviation of the differences: 1.6348834e-07
pytorch_output size: torch.Size([512])
onnx_output size: (1, 512)
```

thanks @fxmarty for the fix, I tried to understand your changes, still on the way to understand more about optimum. Closing the issue.
